### PR TITLE
CC-16646 Bugfix

### PR DESCRIPTION
**Pull request by kkicheglovspryker**
https://spryker.atlassian.net/browse/CC-16646

### CODE REVIEW

File: src/SprykerEco/Zed/Unzer/Business/Checkout/Mapper/UnzerCheckoutMapper.php [hunk 1]
```diff
@@ -106,12 +94,17 @@ protected function mapQuoteItemTransferToUnzerBasketItemTransfer(
         UnzerBasketItemTransfer $unzerBasketItemTransfer
     ): UnzerBasketItemTransfer {
         return $unzerBasketItemTransfer
-            ->setBasketItemReferenceId($itemTransfer->getGroupKey())
+            ->setBasketItemReferenceId(
+                sprintf(
+                    UnzerConstants::UNZER_BASKET_ITEM_REFERENCE_ID_TEMPLATE,
+                    $itemTransfer->getGroupKeyOrFail(),
+                    $itemTransfer->getIdSalesOrderItemOrFail(),
+                ),
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/Unzer/Business/Checkout/Mapper/UnzerCheckoutMapper.php]: I'd consider extracting it to a separate method.

File: src/SprykerEco/Zed/Unzer/Business/Credentials/UnzerCredentialsValidator/UnzerCredentialsStoreRelationsValidator.php [hunk 2]
```diff
@@ -60,7 +60,11 @@ public function validate(UnzerCredentialsTransfer $unzerCredentialsTransfer): Un
         }
 
         $unzerCredentialsConditionsTransfer = (new UnzerCredentialsConditionsTransfer())
-            ->setStoreIds($unzerCredentialsTransfer->getStoreRelationOrFail()->getIdStores());
+            ->setStoreIds($unzerCredentialsTransfer->getStoreRelationOrFail()->getIdStores())
+            ->setTypes([
+                UnzerConstants::UNZER_CONFIG_TYPE_MAIN_MARKETPLACE,
+                UnzerConstants::UNZER_CONFIG_TYPE_STANDARD,
+            ]);
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/Unzer/Business/Credentials/UnzerCredentialsValidator/UnzerCredentialsStoreRelationsValidator.php]: Maybe it'll make sense to extract this array to class const?

File: src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php [hunk 3]
```diff
@@ -65,8 +49,14 @@ protected function mapItemTransferToUnzerRefundItemTransfer(
...
-            ->setAmountGross($itemTransfer->getSumPriceToPayAggregationOrFail() / UnzerConstants::INT_TO_FLOAT_DIVIDER)
+            ->setAmountGross($itemTransfer->getRefundableAmountOrFail() / UnzerConstants::INT_TO_FLOAT_DIVIDER)
             ->setQuantity($itemTransfer->getQuantityOrFail())
-            ->setBasketItemReferenceId($itemTransfer->getGroupKeyOrFail());
+            ->setBasketItemReferenceId(
+                sprintf(
+                    UnzerConstants::UNZER_BASKET_ITEM_REFERENCE_ID_TEMPLATE,
+                    $itemTransfer->getGroupKeyOrFail(),
+                    $itemTransfer->getIdSalesOrderItemOrFail(),
+                ),
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php]: Same here.

File: src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpenseRefundStrategy.php [hunk 4]
```diff
@@ -164,4 +146,45 @@ protected function isPaymentUnzerOrderItemAlreadyRefunded(
...
+     * @param array<\Generated\Shared\Transfer\ItemTransfer> $itemTransfers
+     *
+     * @return bool
+     */
+    protected function allItemsAreRefunded(array $itemTransfers): bool
+    {
+        $refundedItemsCount = 0;
+        foreach ($itemTransfers as $itemTransfer) {
+            if ($itemTransfer->getCanceledAmountOrFail() !== 0) {
+                $refundedItemsCount++;
```
Comment (AsonUnique) [inline src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpenseRefundStrategy.php]: Basically, you don't need to count refunded items here, just return `false` if any item's canceled amount === 0.

**Review by romanhavrylko — APPROVED** [review]

File: src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpenseRefundStrategy.php (hunk 4 above)
Comment (kkicheglovspryker) [reply src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpenseRefundStrategy.php]: I need to check if all items are refunded but not any

**Review by AsonUnique — APPROVED** [review]

File: src/SprykerEco/Zed/Unzer/Business/Checkout/Mapper/UnzerCheckoutMapper.php [hunk 5]
```diff
@@ -106,15 +94,28 @@ protected function mapQuoteItemTransferToUnzerBasketItemTransfer(
...
+    /**
+     * @param \Generated\Shared\Transfer\ItemTransfer $itemTransfer
+     *
+     * @return string
+     */
+    protected function createBasketItemReferenceId(ItemTransfer $itemTransfer): string
+    {
+        return sprintf(
+            UnzerConstants::UNZER_BASKET_ITEM_REFERENCE_ID_TEMPLATE,
+            $itemTransfer->getGroupKeyOrFail(),
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Zed/Unzer/Business/Checkout/Mapper/UnzerCheckoutMapper.php:117]: seems like these fields are not listed as required at API specs

File: src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php [hunk 6]
```diff
@@ -65,8 +49,22 @@ protected function mapItemTransferToUnzerRefundItemTransfer(
...
+
+    /**
+     * @param \Generated\Shared\Transfer\ItemTransfer $itemTransfer
+     *
+     * @return string
+     */
+    protected function createBasketItemReferenceId(ItemTransfer $itemTransfer): string
+    {
+        return sprintf(
+            UnzerConstants::UNZER_BASKET_ITEM_REFERENCE_ID_TEMPLATE,
```
Comment (DmitryLymarenko) [inline src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php:65]: same

**Review by DmitryLymarenko — APPROVED** [review]

**Pull request merged**